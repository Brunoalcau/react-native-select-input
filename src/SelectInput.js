import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  ViewPropTypes,
  Text,
  TouchableWithoutFeedback,
  Keyboard,
  Platform,
  StyleSheet,
} from 'react-native';
import Icon, { ARROW_ICON_TYPES } from './Icon';
import Picker from './Picker';
import Loading from './Loading';

class SelectInput extends Component {
  static propTypes = {
    label: PropTypes.string,
    placeholder: PropTypes.string,
    done: PropTypes.string,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    options: PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.oneOfType([
          PropTypes.string,
          PropTypes.number,
        ]),
        label: PropTypes.string,
      }),
    ).isRequired,
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
    colors: PropTypes.shape({
      normal: PropTypes.string,
      light: PropTypes.string,
      dark: PropTypes.string,
    }).isRequired,
    onChange: PropTypes.func,
    renderLabel: PropTypes.func,
    renderArrowIcon: PropTypes.func,
    labelStyle: Text.propTypes.style,
    valueStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
  };

  static defaultProps = {
    label: null,
    placeholder: 'Select a value',
    done: 'Done',
    value: null,
    arrowIcon: null,
    disabled: false,
    loading: false,
    colors: {
      normal: '#9C9C9C',
      light: '#D9D9D9',
      dark: '#545454',
    },
    onChange: () => {},
    renderLabel: null,
    renderArrowIcon: null,
    labelStyle: null,
    valueStyle: null,
    style: null,
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.value,
      optionsVisible: false,
    };
    this.styles = styles(props);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.state.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  getValue = () => {
    const { options, placeholder } = this.props;

    const selectedOption = options.find(({ value }) => value === this.state.value);

    if (selectedOption) {
      return selectedOption.label;
    }

    return placeholder;
  };

  handleChangeValue = (value) => {
    this.setState({
      value,
    }, () => this.props.onChange(value));
  }

  handleToggleShowOptions = () => {
    if (this.props.disabled || this.props.loading) {
      return;
    }

    Keyboard.dismiss();

    this.setState(prevState => ({
      optionsVisible: !prevState.optionsVisible,
    }));
  };

  renderValue = () => {
    const { loading, disabled, valueStyle } = this.props;

    if (loading) {
      return (
        <View style={this.styles.loadingContainer}>
          <Loading />
        </View>
      );
    }

    return (
      <Text
        style={[
          this.styles.value,
          disabled && this.styles.valueDisabled,
          valueStyle,
        ]}
      >
        {this.getValue()}
      </Text>
    );
  }

  renderLabel = () => {
    const {
      label,
      renderLabel,
      disabled,
      loading,
      labelStyle,
    } = this.props;

    if (renderLabel) {
      return renderLabel();
    }

    if (!label) {
      return;
    }

    return (
      <Text
        style={[
          this.styles.label,
          (disabled || loading) && this.styles.labelDisabled,
          labelStyle,
        ]}
      >
        {label}
      </Text>
    );
  }

  render() {
    const { value, optionsVisible } = this.state;
    const {
      label,
      disabled,
      loading,
      colors,
      style,
      renderArrowIcon,
      ...props
    } = this.props;

    return (
      <View style={[this.styles.container, style]}>
        <TouchableWithoutFeedback onPress={this.handleToggleShowOptions}>
          <View
            style={[
              this.styles.innerContainer,
              this.state.optionsVisible && this.styles.activeInnerContainer,
            ]}
          >
            {this.renderLabel()}

            <View style={this.styles.valueContainer}>
              {this.renderValue()}

              <Icon
                disabled={disabled}
                loading={loading}
                type={optionsVisible ? ARROW_ICON_TYPES.UP : ARROW_ICON_TYPES.DOWN}
                tintColor={(disabled || loading) ? colors.normal : colors.dark}
                renderArrowIcon={renderArrowIcon}
              />
            </View>
          </View>
        </TouchableWithoutFeedback>

        {(!disabled && !loading) && (
          <Picker
            {...props}
            colors={colors}
            value={value}
            onChange={this.handleChangeValue}
            visible={optionsVisible}
            toggleShowOptions={this.handleToggleShowOptions}
          />
        )}
      </View>
    );
  }
}

const styles = ({ colors }) => StyleSheet.create({
  container: {
    zIndex: 1,
    flex: 1,
  },
  innerContainer: {
    flex: 1,
    marginBottom: 28,
    borderBottomWidth: 1,
    borderBottomColor: colors.light,
    paddingVertical: 10,
  },
  activeInnerContainer: {
    borderBottomWidth: 2,
    borderBottomColor: colors.dark,
  },
  label: {
    color: colors.normal,
    fontSize: 14,
  },
  labelDisabled: {
    color: colors.normal,
  },
  valueContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  loadingContainer: {
    paddingTop: Platform.OS === 'ios' ? 9 : 15,
  },
  value: {
    fontSize: 14,
    color: colors.dark,
    paddingTop: Platform.OS === 'ios' ? 9 : 15,
    marginRight: 15,
  },
  valueDisabled: {
    color: colors.normal,
  },
});

export default SelectInput;
