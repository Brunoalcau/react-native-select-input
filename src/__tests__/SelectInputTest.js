import React from 'react';
import { TouchableWithoutFeedback } from 'react-native';
import { shallow } from 'enzyme';
import SelectInput from '../';

describe('SelectInput', () => {
  beforeEach(() => {
    const options = [{
      value: 'first-value',
      label: 'First value',
    }, {
      value: 'second-value',
      label: 'Second value',
    }];

    this.props = {
      label: 'label',
      options,
      value: options[0].value,
      onChange: jest.fn(),
    };
  });

  it('renders correctly', () => {
    const wrapper = shallow(<SelectInput {...this.props} />);

    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with disabled state', () => {
    const wrapper = shallow(
      <SelectInput
        {...this.props}
        disabled
      />,
    );

    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with loading state', () => {
    const wrapper = shallow(
      <SelectInput
        {...this.props}
        loading
      />,
    );

    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with different style', () => {
    const wrapper = shallow(
      <SelectInput
        {...this.props}
        style={{
          backgroundColor: '#0000FF',
        }}
      />,
    );

    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should call onChange prop when the value changes', () => {
    const wrapper = shallow(<SelectInput {...this.props} />);

    const expectedValue = this.props.options[1].value;

    wrapper.instance().handleChangeValue(expectedValue);

    expect(this.props.onChange).toHaveBeenCalledWith(expectedValue);
    expect(wrapper.state().value).toBe(expectedValue);
  });

  it('should open modal with picker', () => {
    const wrapper = shallow(<SelectInput {...this.props} />);

    wrapper.find(TouchableWithoutFeedback).props().onPress();

    expect(wrapper.state().optionsVisible).toBe(true);
  });
});
