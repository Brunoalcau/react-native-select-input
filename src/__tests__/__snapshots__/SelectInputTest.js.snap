// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "zIndex": 1,
      },
      null,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#D9D9D9",
            "borderBottomWidth": 1,
            "flex": 1,
            "marginBottom": 28,
            "paddingVertical": 10,
          },
          false,
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#9C9C9C",
              "fontSize": 14,
            },
            false,
            null,
          ]
        }
      >
        label
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#545454",
                "fontSize": 14,
                "marginRight": 15,
                "paddingTop": 9,
              },
              false,
              null,
            ]
          }
        >
          First value
        </Text>
        <Icon
          disabled={false}
          loading={false}
          renderArrowIcon={null}
          style={null}
          tintColor="#545454"
          type="DOWN"
        />
      </View>
    </View>
  </TouchableWithoutFeedback>
  <Picker
    arrowIcon={null}
    colors={
      Object {
        "dark": "#545454",
        "light": "#D9D9D9",
        "normal": "#9C9C9C",
      }
    }
    done="Done"
    doneStyle={null}
    labelStyle={null}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "First value",
          "value": "first-value",
        },
        Object {
          "label": "Second value",
          "value": "second-value",
        },
      ]
    }
    placeholder="Select a value"
    renderLabel={null}
    testProperty={null}
    toggleShowOptions={[Function]}
    value="first-value"
    valueStyle={null}
    visible={false}
  />
</View>
`;

exports[`SelectInput renders correctly with a testProperty set 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "zIndex": 1,
      },
      null,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
    testID="select-foo"
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#D9D9D9",
            "borderBottomWidth": 1,
            "flex": 1,
            "marginBottom": 28,
            "paddingVertical": 10,
          },
          false,
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#9C9C9C",
              "fontSize": 14,
            },
            false,
            null,
          ]
        }
      >
        label
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#545454",
                "fontSize": 14,
                "marginRight": 15,
                "paddingTop": 9,
              },
              false,
              null,
            ]
          }
        >
          First value
        </Text>
        <Icon
          disabled={false}
          loading={false}
          renderArrowIcon={null}
          style={null}
          tintColor="#545454"
          type="DOWN"
        />
      </View>
    </View>
  </TouchableWithoutFeedback>
  <Picker
    arrowIcon={null}
    colors={
      Object {
        "dark": "#545454",
        "light": "#D9D9D9",
        "normal": "#9C9C9C",
      }
    }
    done="Done"
    doneStyle={null}
    labelStyle={null}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "First value",
          "value": "first-value",
        },
        Object {
          "label": "Second value",
          "value": "second-value",
        },
      ]
    }
    placeholder="Select a value"
    renderLabel={null}
    testProperty="foo"
    toggleShowOptions={[Function]}
    value="first-value"
    valueStyle={null}
    visible={false}
  />
</View>
`;

exports[`SelectInput renders correctly with different style 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "zIndex": 1,
      },
      null,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#D9D9D9",
            "borderBottomWidth": 1,
            "flex": 1,
            "marginBottom": 28,
            "paddingVertical": 10,
          },
          false,
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#9C9C9C",
              "fontSize": 14,
            },
            false,
            null,
          ]
        }
      >
        label
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#545454",
                "fontSize": 14,
                "marginRight": 15,
                "paddingTop": 9,
              },
              false,
              null,
            ]
          }
        >
          First value
        </Text>
        <Icon
          disabled={false}
          loading={false}
          renderArrowIcon={null}
          style={null}
          tintColor="#545454"
          type="DOWN"
        />
      </View>
    </View>
  </TouchableWithoutFeedback>
  <Picker
    arrowIcon={null}
    colors={
      Object {
        "dark": "#545454",
        "light": "#D9D9D9",
        "normal": "#9C9C9C",
      }
    }
    done="Done"
    doneStyle={null}
    labelStyle={null}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "First value",
          "value": "first-value",
        },
        Object {
          "label": "Second value",
          "value": "second-value",
        },
      ]
    }
    placeholder="Select a value"
    renderLabel={null}
    style={
      Object {
        "backgroundColor": "#0000FF",
      }
    }
    testProperty={null}
    toggleShowOptions={[Function]}
    value="first-value"
    valueStyle={null}
    visible={false}
  />
</View>
`;

exports[`SelectInput renders correctly with disabled state 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "zIndex": 1,
      },
      null,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#D9D9D9",
            "borderBottomWidth": 1,
            "flex": 1,
            "marginBottom": 28,
            "paddingVertical": 10,
          },
          false,
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#9C9C9C",
              "fontSize": 14,
            },
            Object {
              "color": "#9C9C9C",
            },
            null,
          ]
        }
      >
        label
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#545454",
                "fontSize": 14,
                "marginRight": 15,
                "paddingTop": 9,
              },
              Object {
                "color": "#9C9C9C",
              },
              null,
            ]
          }
        >
          First value
        </Text>
        <Icon
          disabled={true}
          loading={false}
          renderArrowIcon={null}
          style={null}
          tintColor="#9C9C9C"
          type="DOWN"
        />
      </View>
    </View>
  </TouchableWithoutFeedback>
</View>
`;

exports[`SelectInput renders correctly with loading state 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "zIndex": 1,
      },
      null,
    ]
  }
>
  <TouchableWithoutFeedback
    onPress={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "borderBottomColor": "#D9D9D9",
            "borderBottomWidth": 1,
            "flex": 1,
            "marginBottom": 28,
            "paddingVertical": 10,
          },
          false,
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#9C9C9C",
              "fontSize": 14,
            },
            Object {
              "color": "#9C9C9C",
            },
            null,
          ]
        }
      >
        label
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Object {
              "paddingTop": 9,
            }
          }
        >
          <Loading
            color="#000"
            size="small"
          />
        </View>
        <Icon
          disabled={false}
          loading={true}
          renderArrowIcon={null}
          style={null}
          tintColor="#9C9C9C"
          type="DOWN"
        />
      </View>
    </View>
  </TouchableWithoutFeedback>
</View>
`;
